core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage2.java

```diff
@@ -0,0 +1,155 @@
+import software.amazon.awssdk.core.internal.http.TransformingAsyncResponseHandler;
+import software.amazon.awssdk.core.internal.http.pipeline.RequestPipeline;
+import software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper2;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+
+/**
+ * Wrapper around the pipeline for a single request to provide retry, clockskew and request throttling functionality.
+ */
+@SdkInternalApi
+public final class AsyncRetryableStage2<OutputT> implements RequestPipeline<SdkHttpFullRequest,
+    CompletableFuture<Response<OutputT>>> {
+
+    private final TransformingAsyncResponseHandler<Response<OutputT>> responseHandler;
+    private final RequestPipeline<SdkHttpFullRequest, CompletableFuture<Response<OutputT>>> requestPipeline;
+    private final ScheduledExecutorService scheduledExecutor;
+    private final HttpClientDependencies dependencies;
+
+    public AsyncRetryableStage2(TransformingAsyncResponseHandler<Response<OutputT>> responseHandler,
+                                HttpClientDependencies dependencies,
+                                RequestPipeline<SdkHttpFullRequest, CompletableFuture<Response<OutputT>>> requestPipeline) {
+        this.responseHandler = responseHandler;
+        this.dependencies = dependencies;
+        this.scheduledExecutor = dependencies.clientConfiguration().option(SdkClientOption.SCHEDULED_EXECUTOR_SERVICE);
+        this.requestPipeline = requestPipeline;
+    }
+
+    @Override
+    public CompletableFuture<Response<OutputT>> execute(SdkHttpFullRequest request,
+                                                        RequestExecutionContext context) throws Exception {
+        return new RetryingExecutor(request, context).execute();
+    }
+
+    private final class RetryingExecutor {
+        private final AsyncRequestBody originalRequestBody;
+        private final RequestExecutionContext context;
+        private final RetryableStageHelper2 retryableStageHelper;
+
+        private RetryingExecutor(SdkHttpFullRequest request, RequestExecutionContext context) {
+            this.originalRequestBody = context.requestProvider();
+            this.context = context;
+            this.retryableStageHelper = new RetryableStageHelper2(request, context, dependencies);
+        }
+
+        public CompletableFuture<Response<OutputT>> execute() {
+            CompletableFuture<Response<OutputT>> future = new CompletableFuture<>();
+            attemptFirstExecute(future);
+            return future;
+        }
+
+        public void attemptFirstExecute(CompletableFuture<Response<OutputT>> future) {
+            Duration backoffDelay = retryableStageHelper.acquireInitialToken();
+            if (backoffDelay.isZero()) {
+                attemptExecute(future);
+            } else {
+                retryableStageHelper.logBackingOff(backoffDelay);
```

Comment:
I could not find a Junit for this case , how is this tested ?

Reply:
I will add more tests for these after this is merged in a follow up PR; the idea is to make smaller changes to make it easier to review with self-contained minimal changes that build on top of each other.